### PR TITLE
 moved code into class

### PR DESCRIPTION
**Pull request by TheBigFatTony**
moved code into class
added unit tests
can now use rbg-array as base-palette
added new functions: convert_palette_to_rbg
changed README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -63,8 +63,7 @@ usage: glasbey.py [-h] [--base-palette BASE_PALETTE] [--no-black] [--view]
 
     If having black (and colors close to black) is undesired, then `--no-black`
     option may be used to prevent the algorithm from inserting such colors into
-    the palette. In addition to that, the range of colors considered for
-    inclusion in the palette can be limited by lightness, chroma, or hue.
+    the palette.
```
Comment (taketwo) [inline README.md]: Some documentation got lost, could you please add it back? Also positional arguments below.

File: glasbey.py [hunk 2]
```diff
@@ -1,29 +1,36 @@
 #!/usr/bin/env python
 # encoding: utf-8
-
-import sys
+import os, sys
```
Comment (taketwo) [inline glasbey.py]: Please re-add empty line before and put one import per line.

File: glasbey.py [hunk 3]
```diff
@@ -38,118 +45,227 @@ def finish(self):
...
-    if hue_range is not None:
-        jch = cspace_convert(colors, "CAM02-UCS", "JCh")
-        if hue_range[0] > hue_range[1]:
-            colors = colors[
-                (jch[:, 2] >= hue_range[0]) | (jch[:, 2] <= hue_range[1]), :
-            ]
+class Glasbey:
+    def __init__(self,
+                 base_palette=None,
+                 overwrite_base_palette: bool = False,
```
Comment (taketwo) [inline glasbey.py]: What's the point of overwriting base palette?

File: glasbey.py [hunk 4]
```diff
@@ -38,118 +45,227 @@ def finish(self):
...
+                self.colors = self.colors[
+                              (jch[:, 2] >= hue_range[0]) & (jch[:, 2] <= hue_range[1]), :
+                              ]
+        # Exclude colors that are close to black
+        if no_black:
+            MIN_DISTANCE_TO_BLACK = 35
+            d = np.linalg.norm((self.colors - self.colors[0, :]), axis=1)
+            self.colors = self.colors[d > MIN_DISTANCE_TO_BLACK, :]
+
+    def get_palette(self, size):
```
Comment (taketwo) [inline glasbey.py]: What about calling this `generate_palette`? Hints at the fact that there may be computation involved.
Also, I would merge `extend_palette` into this function. Something like:

```python
if size <= len(self.palette):
    return cspace_convert(...)
# extend palette code
```

File: glasbey.py [hunk 5]
```diff
@@ -38,118 +45,227 @@ def finish(self):
...
+
+        If the palette isn't long enough, new entries are generated.
+        """
+
+        if size > len(self.palette):
+            # need to extend palette
+            self.extend_palette(size=size)
+        return cspace_convert(self.palette[0:size], "CAM02-UCS", "sRGB1")
+
+    def load_colors(self):
```
Comment (taketwo) [inline glasbey.py]: I'd call this `load_or_generate_color_table()`.

File: glasbey.py [hunk 6]
```diff
@@ -38,118 +45,227 @@ def finish(self):
...
+            self.palette.append(self.colors[np.argmax(distances), :])
+            pbar.update(len(self.palette))
+        pbar.finish()
+
+        assert self.palette_invariant(), "Internal error during extend_palette: self.palette is poorly formatted."
+
+        if self.overwrite_base_palette:
+            self.save_palette(palette=self.palette, path=self.base_palette, format="byte", overwrite=True)
+
+    def load_palette(self, path):
```
Comment (taketwo) [inline glasbey.py]: Can be static.

File: glasbey.py [hunk 7]
```diff
@@ -38,118 +45,227 @@ def finish(self):
...
+                for color in palette:
+                    rgb255 = tuple(int(round(k * 255)) for k in color)
+                    file.write("{},{},{}\n".format(*rgb255))
+            elif format.lower() == "float":
+                for color in palette:
+                    file.write("{:.6f},{:.6f},{:.6f}\n".format(*(abs(k) for k in color)))
+            else:
+                raise ValueError("Format doesn't match. Choose between 'byte' and 'float'")
+
+    def palette_invariant(self):
```
Comment (taketwo) [inline glasbey.py]: Are you sure "invariant" is what you meant? For me rather looks like the function is answering question "is palette valid?"

Comment (TheBigFatTony) [reply glasbey.py]: Renamed the functions to...
check_validity_rbg_palette()
check_validity_internal_palette()

File: glasbey.py (hunk 6 above)
Comment (TheBigFatTony) [reply glasbey.py]: Done.

File: glasbey.py (hunk 5 above)
Comment (TheBigFatTony) [reply glasbey.py]: Done.

File: glasbey.py (hunk 4 above)
Comment (TheBigFatTony) [reply glasbey.py]: Done.

File: glasbey.py (hunk 3 above)
Comment (TheBigFatTony) [reply glasbey.py]: I found it useful in my use case, where my code may iteratively generates additional colors.

The parameter is optional and the feature takes only about 3 lines of code.

The advantage is that Glasbey saves the new colors in a format and place it understands perfectly. (Good responsibility-driven design, no?)

File: glasbey.py (hunk 2 above)
Comment (TheBigFatTony) [reply glasbey.py]: Done.

File: glasbey.py (hunk 3 above)
Comment (taketwo) [reply glasbey.py]: I'm fine having this option, though I'm still not sure if I understand your motivation/use-case :)

**Pull request closed without merging**